File: SRC/CPCDOS EDITOR/TabOrderHooker.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.ComponentModel.Design
Imports System.Windows.Forms
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Reflection
Imports System.Diagnostics

Namespace pF.DesignSurfaceExt
    Public Class TabOrderHooker
        Private Const _Name_ As String = "TabOrderHooker"
        Private _tabOrder As Object = Nothing

        Public Sub HookTabOrder(ByVal host As IDesignerHost)
            Const _signature_ As String = _Name_ & "::ctor()"
            If host.RootComponent Is Nothing Then Throw New Exception(_signature_ & " - Exception: the TabOrder must be invoked after the DesignSurface has been loaded!")

            Try
                Dim designAssembly As System.Reflection.Assembly = System.Reflection.Assembly.Load("System.Design, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a")
                Dim tabOrderType As Type = designAssembly.[GetType]("System.Windows.Forms.Design.TabOrder")

                If _tabOrder Is Nothing Then
                    _tabOrder = Activator.CreateInstance(tabOrderType, New Object() {host})
                Else
                    DisposeTabOrder()
                End If

            Catch exx As Exception
                Debug.WriteLine(exx.Message)
                If exx.InnerException IsNot Nothing Then Debug.WriteLine(exx.InnerException.Message)
                Throw
            End Try
        End Sub

        Public Sub DisposeTabOrder()
            If _tabOrder Is Nothing Then Return

            Try
                Dim designAssembly As System.Reflection.Assembly = System.Reflection.Assembly.Load("System.Design, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a")
                Dim tabOrderType As Type = designAssembly.[GetType]("System.Windows.Forms.Design.TabOrder")
                tabOrderType.InvokeMember("Dispose", BindingFlags.[Public] Or BindingFlags.Instance Or BindingFlags.NonPublic Or BindingFlags.InvokeMethod, Nothing, _tabOrder, New Object() {True})
                _tabOrder = Nothing
            Catch exx As Exception
                Debug.WriteLine(exx.Message)
                If exx.InnerException IsNot Nothing Then Debug.WriteLine(exx.InnerException.Message)
                Throw
            End Try
        End Sub
    End Class
End Namespace

File: SRC/CPCDOS EDITOR/Classes/TabOrderHooker.cs
using System;
using System.ComponentModel.Design;
using System.Diagnostics;
using System.Reflection;

namespace OSMaker.pF.DesignSurfaceExt
{
    public class TabOrderHooker
    {
        private const string _Name_ = "TabOrderHooker";
        private object _tabOrder = null;

        public void HookTabOrder(IDesignerHost host)
        {
            const string _signature_ = _Name_ + "::ctor()";
            if (host.RootComponent is null)
                throw new Exception(_signature_ + " - Exception: the TabOrder must be invoked after the DesignSurface has been loaded!");
            try
            {
                var designAssembly = Assembly.Load("System.Design, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a");
                var tabOrderType = designAssembly.GetType("System.Windows.Forms.Design.TabOrder");
                if (_tabOrder is null)
                {
                    _tabOrder = Activator.CreateInstance(tabOrderType, new object[] { host });
                }
                else
                {
                    DisposeTabOrder();
                }
            }
            catch (Exception exx)
            {
                Debug.WriteLine(exx.Message);
                if (exx.InnerException is object)
                    Debug.WriteLine(exx.InnerException.Message);
                throw;
            }
        }

        public void DisposeTabOrder()
        {
            if (_tabOrder is null)
                return;
            try
            {
                var designAssembly = Assembly.Load("System.Design, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a");
                var tabOrderType = designAssembly.GetType("System.Windows.Forms.Design.TabOrder");
                tabOrderType.InvokeMember("Dispose", BindingFlags.Public | BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.InvokeMethod, null, _tabOrder, new object[] { true });
                _tabOrder = null;
            }
            catch (Exception exx)
            {
                Debug.WriteLine(exx.Message);
                if (exx.InnerException is object)
                    Debug.WriteLine(exx.InnerException.Message);
                throw;
            }
        }
    }
}